### DISCUSSION TITLE
Branching strategy

### DISCUSSION THREAD

**Discussion by lukas-reineke** [famiu/feline.nvim#187] [General]
I have some suggestions regarding the branching strategy.
Mainly about how `develop` is used.
`develop` is a long-lived branch. And users might want to use this version. It should not be rebased / force-pushed.

I am all for requiring rebase and linear history for PRs against `develop`. But `develop` itself should have a stable history.
This makes developing with multiply people much easier.
It does not cause issues when users use the branch.
We can use GitHub settings to properly protect it and setup rules for it.

So TLDR;
`feature branch -> develop` = rebase + linear history
`develop -> master` = merge commit
both develop and master are protected, and the history does not change.

**Comment by famiu** [comment]
I actually agree and I think we could adopt that approach, so no complaints from me

**Comment by lukas-reineke** [comment]
@famiu can you protect the develop branch then? I don't have access to do that.
You can also check `Require linear history` for develop. So no one accidentally makes merge commits like I did.
And should we merge #185  `Require status checks to pass before merging` and `Require branches to be up to date before merging` would also be nice, for both develop and master.

**Comment by famiu** [comment]
Do you think master itself should also require linear history?

**Comment by lukas-reineke** [comment]
Requiring linear history on master and develop doesn't work. Because you want to merge them into each other.
They can't be both linear, except when you rebase, but we want to avoid that because it rewrites history.

In practice, there shouldn't be many cases where you need to hotfix PRs directly into master.
For me, just having develop be forced linear, and being careful with master, was always good enough.

**Reply by famiu** [reply]
> Requiring linear history on master and develop doesn't work. Because you want to merge them into each other. They can't be both linear, except when you rebase, but we want to avoid that because it rewrites history.
> 
> In practice, there shouldn't be many cases where you need to hotfix PRs directly into master. For me, just having develop be forced linear, and being careful with master, was always good enough.

Ah, that makes sense. Alright then. I've done what was required. I'm just wondering though if it would be possible to only allow merging develop to master and restrict merging from any other branch / fork to master to prevent accidental merges.

**Reply by lukas-reineke** [reply]
> I'm just wondering though if it would be possible to only allow merging develop to master and restrict merging from any other branch / fork to master to prevent accidental merges.

Would be nice, but I don't think GitHub can do that.
You can require PR reviews before merging, that might help to make sure at least 2 people take a look
